keps/sig-autoscaling/20181106-in-place-update-of-pod-resources.md

```diff
@@ -0,0 +1,375 @@
+Node that accommodates the Pod.
+
+For a newly created Pod, Spec.Containers[i].ResourcesAllocated must match
+Spec.Containers[i].Resources.Requests. When Kubelet admits a new Pod, values in
+Spec.Containers[i].ResourcesAllocated are used to determine if there is enough
+room to admit the Pod. Kubelet does not set Pod's ResourcesAllocated after
+admitting a new Pod.
+
+When a Pod resize is requested, Kubelet attempts to update the resources
+allocated to the Pod and its Containers. Kubelet first checks if the new
+desired resources can fit the Node allocable resources by computing the sum of
+resources allocated (Pod.Spec.Containers[i].ResourcesAllocated) for all Pods in
+the Node, except the Pod being resized. For the Pod being resized, it adds the
+new desired resources (i.e Spec.Containers[i].Resources.Requests) to the sum.
+* If new desired resources fit, Kubelet accepts the resize by updating
+  Pod.Spec.Containers[i].ResourcesAllocated via pods/resourceallocation
+  subresource, and then proceeds to invoke UpdateContainerResources CRI API
+  to update the Container resource limits. Once all Containers are successfully
+  updated, it updates Pod.Status.ContainerStatuses[i].Resources to reflect the
+  new resource values.
+* If new desired resources don't fit, Kubelet rejects the resize, and no
+  further action is taken.
+  - Kubelet retries the Pod resize at a later time.
+
+If multiple Pods need resizing, they are handled sequentially in the order in
+which Pod additions and updates arrive at Kubelet.
+
+Scheduler may, in parallel, assign a new Pod to the Node because it uses cached
+Pods to compute Node allocable values. If this race condition occurs, Kubelet
+resolves it by rejecting that new Pod if the Node has no room after Pod resize.
+
+#### Kubelet Restart Tolerance
+
+If Kubelet were to restart amidst handling a Pod resize, then upon restart, all
+Pods are admitted at their current Pod.Spec.Containers[i].ResourcesAllocated
+values, and resizes are handled after all existing Pods have been added. This
+ensures that resizes don't affect previously admitted existing Pods.
+
+### Scheduler and API Server Interaction
+
+Scheduler continues to use Pod's Spec.Containers[i].Resources.Requests for
+scheduling new Pods, and continues to watch Pod updates, and updates its cache.
+It uses the cached Pod's Spec.Containers[i].ResourcesAllocated values to
+compute the Node resources allocated to Pods. This ensures that it always uses
+the most recently available resource allocations in making new Pod scheduling
+decisions.
+
+### Flow Control
+
+The following steps denote a typical flow of an in-place resize operation for a
+Pod with ResizePolicy set to NoRestart for all its Containers.
+
+1. Initiating actor updates Pod's Spec.Containers[i].Resources via PATCH verb.
+1. API Server validates the new Resources. (e.g. Limits are not below
+   Requests, QoS class doesn't change, ResourceQuota not exceeded...)
```

Comment:
Say that a pod started out with requests lower than its limits, giving it QoS class "Burstable," and we update it later increasing the requests to match the original limits. Had it started that way, it would have gotten QoS class "Guaranteed."

Does this statement above mean that the QoS class stays as it was—at "Burstable"—even with the change in requests, or that the API server would reject this change, because it would force a change to the QoS class?

Reply:
This will be rejected in APIserver validation.

Reply:
To be sure I understand, let's use an example. I create a pod with the following resources:
```yaml
# ...
spec:
  containers:
  - resources:
      requests:
        cpu: 50m
        memory: 250Mi
```
Later, I submit an update with the following:
```yaml
# ...
spec:
  containers:
  - resources:
      requests:
        cpu: 50m
        memory: 250Mi
      limits:
        cpu: 50m
        memory: 250Mi
```
Are you saying the latter would be rejected, because the QoS class would change?

If so, how about this update?
```yaml
# ...
spec:
  containers:
  - resources:
      requests:
        cpu: 50m
        memory: 250Mi
      limits:
        cpu: 50m
        memory: 251Mi  # NB: It's one higher than the request.
```

Reply:
To be clear, I understand that clients can't submit initial values for nor changes to the pod's QoS class explicitly. What I'm wondering, though, is two things:
1. Will a pod's QoS class ever change after creation?
2. If It won't change, will we allow resource request and limit changes that render the initially assessed QoS class no longer accurate?

Reply:
Current implementation manages pod level resources using linux cgroups, and kubelet puts burstable and besteffort class pods under paths named as such - e.g. /sys/fs/cgroup/memory/kubepods/burstable/podabc... 

Changing QoS class means allowing cgroup path change. Without further analysis as to its feasibility and implications, I cannot say if it can be done. For example, is it possible to move a running pod's location in the cgroup hierarchy without stopping? Can this property, if available, be used by pods to game the system, work around eviction algorithm etc.?

This is outside the scope of the current KEP, hence we keep things simple and reject it in validation. You can change requests and/or limits as long as the QoS class does not change. In above example, the first change is disallowed, second change would be allowed since you're going from no-limits to imposing limits within the burstable QoS class.

Reply:
Thank you very much for the thorough explanation. The situation—and the challenges—are clear to me now.